drivers/clock_control/clock_control_nrfs_audiopll.c

```diff
@@ -0,0 +1,211 @@
+ */
+
+#define DT_DRV_COMPAT nordic_nrfs_audiopll
+
+#include <zephyr/devicetree.h>
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/drivers/clock_control/nrf_clock_control.h>
+
+#include <nrfs_audiopll.h>
+#include <nrfs_backend_ipc_service.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(nrfs_audiopll, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
+
+BUILD_ASSERT(
+	DT_NUM_INST_STATUS_OKAY(DT_DRV_COMPAT) == 1,
+	"multiple instances not supported"
+);
+
+BUILD_ASSERT((int)AUDIOPLL_DIV_DISABLED == (int)NRF_CLOCK_CONTROL_AUDIOPLL_DIV_DISABLED);
+BUILD_ASSERT((int)AUDIOPLL_DIV_1 == (int)NRF_CLOCK_CONTROL_AUDIOPLL_DIV_1);
+BUILD_ASSERT((int)AUDIOPLL_DIV_2 == (int)NRF_CLOCK_CONTROL_AUDIOPLL_DIV_2);
+BUILD_ASSERT((int)AUDIOPLL_DIV_3 == (int)NRF_CLOCK_CONTROL_AUDIOPLL_DIV_3);
+BUILD_ASSERT((int)AUDIOPLL_DIV_4 == (int)NRF_CLOCK_CONTROL_AUDIOPLL_DIV_4);
+BUILD_ASSERT((int)AUDIOPLL_DIV_6 == (int)NRF_CLOCK_CONTROL_AUDIOPLL_DIV_6);
+BUILD_ASSERT((int)AUDIOPLL_DIV_8 == (int)NRF_CLOCK_CONTROL_AUDIOPLL_DIV_8);
+BUILD_ASSERT((int)AUDIOPLL_DIV_12 == (int)NRF_CLOCK_CONTROL_AUDIOPLL_DIV_12);
+BUILD_ASSERT((int)AUDIOPLL_DIV_16 == (int)NRF_CLOCK_CONTROL_AUDIOPLL_DIV_16);
+
+struct shim_data {
+	const struct device *dev;
+	struct k_sem evt_sem;
+	nrfs_audiopll_evt_type_t evt;
+};
+
+struct shim_config {
+	uint16_t freq_fraction;
+	enum nrf_clock_control_audiopll_prescaler_div div;
+};
+
+static void shim_nrfs_audiopll_evt_handler(nrfs_audiopll_evt_t const *evt, void *context)
+{
+	struct shim_data *dev_data = context;
+
+	dev_data->evt = evt->type;
+	k_sem_give(&dev_data->evt_sem);
+}
+
+static int shim_nrfs_request_enable(const struct device *dev)
+{
+	struct shim_data *dev_data = dev->data;
+	nrfs_err_t err;
+
+	LOG_DBG("send enable request");
+
+	err = nrfs_audiopll_enable_request(dev_data);
+	if (err != NRFS_SUCCESS) {
+		return -EAGAIN;
+	}
+
+	k_sem_take(&dev_data->evt_sem, K_FOREVER);
+	return dev_data->evt == NRFS_AUDIOPLL_EVT_ENABLED ? 0 : -EAGAIN;
+}
+
+static int shim_nrfs_request_disable(const struct device *dev)
+{
+	struct shim_data *dev_data = dev->data;
+	nrfs_err_t err;
+
+	LOG_DBG("send disable request");
+
+	err = nrfs_audiopll_disable_request(dev_data);
+	if (err != NRFS_SUCCESS) {
+		return -EAGAIN;
+	}
+
+	k_sem_take(&dev_data->evt_sem, K_FOREVER);
+	return dev_data->evt == NRFS_AUDIOPLL_EVT_DISABLED ? 0 : -EAGAIN;
+}
+
+static int shim_clock_control_api_on(const struct device *dev,
+				     clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(sub_system);
+
+	return shim_nrfs_request_enable(dev);
+}
+
+static int shim_clock_control_api_off(const struct device *dev,
+				      clock_control_subsys_t sub_system)
+{
+	ARG_UNUSED(sub_system);
+
+	return shim_nrfs_request_disable(dev);
+}
+
+static DEVICE_API(clock_control, shim_clock_control_api) = {
+	.on = shim_clock_control_api_on,
+	.off = shim_clock_control_api_off,
+};
+
+int nrf_clock_control_audiopll_set_freq(const struct device *dev, uint16_t freq_fraction)
```

Comment:
From the application layer, I hope this could be simplified to just setting the frequency we want directly, and not have to consider `freq_fraction `and `prescaler`.
This change will also be different to the current PLL API where we just set frequency.

Reply:
Replaced direct API with a simple set/get frequency, using the clock_control API